infer/internal/ende/ende.go

```diff
@@ -300,9 +315,38 @@ func (e *ende) Encode(src any) (resource.PropertyMap, mapper.MappingError) {
 	if err != nil {
 		return nil, err
 	}
-	m := resource.NewObjectProperty(
-		resource.NewPropertyMapFromMap(props),
-	)
+
+	// If we see the magic signatures meaning "asset" or "archive", it's an AssetOrArchive and need
+	// to pull the actual, inner asset or archive out of the object and discard the outer
+	// AssetOrArchive. See #237 for more background.
+	// The literal magic signatures are from pulumi/pulumi and are not exported by the SDK.
+	m := resource.NewPropertyValueRepl(props,
+		nil, // keys are not changed
+		func(a any) (resource.PropertyValue, bool) {
+			if aMap, ok := a.(map[string]any); ok {
+				if rawAsset, ok := aMap[types.AssetSignature]; ok {
+					if asset, ok := rawAsset.(map[string]any); ok {
+						if sig, ok := asset["4dabf18193072939515e22adb298388d"]; ok {
+							if sigStr, ok := sig.(string); ok && sigStr == "c44067f5952c0a294b673a41bacd8c17" {
```

Comment:
Can we use the [constant in `pulumi/pulumi`](https://github.com/pulumi/pulumi/blob/5a83cfe6e4b77705240499134fae29d68c3c3ddf/sdk/go/common/resource/sig/sig.go#L18-L21)?
```suggestion
						if kind, ok := asset[sig.Key]; ok {
							if kind, ok := kind.(string); ok && kind == sig.AssetSig {

```